src/java/org/apache/cassandra/index/sai/utils/RangeConcatIterator.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.utils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.PriorityQueue;
+
+import org.apache.cassandra.io.util.FileUtils;
+
+/**
+ * {@link RangeConcatIterator} takes a list of sorted range iterator and concatenates them, leaving duplicates in
+ * place, to produce a new stably sorted iterator. Duplicates are eliminated later in
+ * {@link org.apache.cassandra.index.sai.plan.StorageAttachedIndexSearcher}
+ * as results from multiple SSTable indexes and their respective segments are consumed.
+ *
+ * ex. (1, 2, 3) + (3, 3, 4, 5) -> (1, 2, 3, 3, 3, 4, 5)
+ * ex. (1, 2, 2, 3) + (3, 4, 4, 6, 6, 7) -> (1, 2, 2, 3, 3, 4, 4, 6, 6, 7)
+ *
+ */
+public class RangeConcatIterator extends RangeIterator
+{
+    private final PriorityQueue<RangeIterator> ranges;
+    private final List<RangeIterator> toRelease;
+
+    protected RangeConcatIterator(RangeIterator.Builder.Statistics statistics, PriorityQueue<RangeIterator> ranges)
+    {
+        super(statistics);
+
+        this.ranges = ranges;
+        this.toRelease = new ArrayList<>(ranges);
+    }
+
+    @Override
+    @SuppressWarnings("resource")
+    protected void performSkipTo(PrimaryKey primaryKey)
+    {
+        while (!ranges.isEmpty())
+        {
+            if (ranges.peek().getCurrent().compareTo(primaryKey) >= 0)
+                break;
+
+            RangeIterator head = ranges.poll();
+
+            if (head.getMaximum().compareTo(primaryKey) >= 0)
+            {
+                head.skipTo(primaryKey);
+                ranges.add(head);
+                break;
+            }
```

Comment:
What if we changed this (assuming we keep the PQ, see above) to peek...

```
RangeIterator head = ranges.peek();

if (head.getMaximum().compareTo(primaryKey) >= 0)
{
    head.skipTo(primaryKey);
    break;
}
else
    ranges.poll();
```
...or...
```
RangeIterator head = ranges.peek();

if (head.getMaximum().compareTo(primaryKey) < 0)
    ranges.poll();
else
{
    head.skipTo(primaryKey);
    break;
}
```
...to avoid having to add the `head` back? (It doesn't look like anything in `RangeConcatIteratorTest` fails if we do.)

Reply:
hmmm, would there be a problem w/ a data model like `testOverlappingBoundaries()`? Either way, might be good to test skip w/ some overlapping boundaries.

Reply:
Also took a stab at just replacing the PQ here: https://github.com/maedhroz/cassandra/pull/6/files#r1053678339